### PR TITLE
fix: Prevent Authz MsgExec bypass Interchain Security Disabled Modules Messages

### PR DESCRIPTION
**Pull request by sontrinh16**
## Description

Closes: #1492 

---

## Changes
- Add `nestedAuthzMsgExecCheck` func for filtering `/cosmos.authz.v1beta1.MsgExec` msgs and check all the wrapped msgs with `hasDisabledModuleMsgs` func. If there is a MsgExec among the wrapped msgs, repeat the process again until all the wrapped msgs is check or when found a disabled module msgs.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [x] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [x] Added `!` to the type prefix if state-machine breaking change (i.e., requires coordinated upgrade)
* [x] Confirmed this PR does not introduce changes requiring state migrations, OR migration code has been added to consumer and/or provider modules
* [x] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
* [x] Provided a link to the relevant issue or specification
* [x] Followed the guidelines for [building SDK modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/00-intro.md)
* [x] Included the necessary unit and integration [tests](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#testing)
* [ ] Added a changelog entry to `CHANGELOG.md`
* [ ] Included comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] Updated the relevant documentation or specification
* [x] Reviewed "Files changed" and left comments if necessary 
* [x] Confirmed all CI checks have passed
* [ ] If this PR is library API breaking, bump the go.mod version string of the repo, and follow through on a new major release

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [x] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [x] confirmed `!` in the type prefix if API or client breaking change
* [ ] confirmed this PR does not introduce changes requiring state migrations, OR confirmed migration code has been added to consumer and/or provider modules
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic
* [ ] reviewed API design and naming
* [ ] reviewed documentation is accurate
* [ ] reviewed tests and test coverage

### CODE REVIEW

**Comment by p-offtermatt** [comment]
I think this change requires a coordinated upgrade, right? Otherwise, some nodes may filter out the newly disabled types, while other nodes would still process them. Does that seem right to you?

**Comment by sontrinh16** [comment]
> I think this change requires a coordinated upgrade, right? Otherwise, some nodes may filter out the newly disabled types, while other nodes would still process them. Does that seem right to you?

yesss i think so

File: app/consumer/ante_handler.go [hunk 1]
```diff
@@ -42,7 +42,7 @@ func NewAnteHandler(options HandlerOptions) (sdk.AnteHandler, error) {
 	anteDecorators := []sdk.AnteDecorator{
 		ante.NewSetUpContextDecorator(),
 		ante.NewExtensionOptionsDecorator(nil),
-		consumerante.NewMsgFilterDecorator(options.ConsumerKeeper),
+		consumerante.NewMsgFilterDecorator(options.ConsumerKeeper, "/cosmos.evidence", "/cosmos.slashing"),
```
Comment (mpoke) [inline app/consumer/ante_handler.go:45]: Adding the filtering logic in the `MsgFilterDecorator` would require all the consumer chains to update their application, i.e., adding this line to their `ante_handler.go`. An alternative is to add the logic directly to the `DisabledModulesDecorator`, which would require consumer chains to only upgrade their version of ICS.

File: app/consumer/ante/msg_filter_ante.go [hunk 2]
```diff
@@ -54,3 +64,43 @@ func hasValidMsgsPreCCV(msgs []sdk.Msg) bool {
...
+
+			wrappedMsgs = append(wrappedMsgs, sdkMsgs...)
+
+		}
+	}
+
+	return wrappedMsgs
+}
+
+func hasDisabledModuleMsgs(msgs []sdk.Msg, prefixes ...string) bool {
```
Comment (mpoke) [inline app/consumer/ante/msg_filter_ante.go:94]: Moving this logic to the `DisabledModulesDecorator` would also avoid duplicating this code as it's already there.

File: app/consumer/ante_handler.go (hunk 1 above)
Comment (sontrinh16) [reply app/consumer/ante_handler.go:45]: hmmm make sense, i will move the logic to `DisabledModulesDecorator`

Comment (sontrinh16) [reply app/consumer/ante_handler.go]: updated as request

**Review by mpoke — COMMENTED** [review]
Thanks @sontrinh16. See my comments bellow (nits). My only concern is about nested authz exec messages. For example, the following message should be filtered out, but it is not:
```go
	authzMsgExecEvidence := authz.NewMsgExec(sdk.AccAddress{}, []sdk.Msg{&evidencetypes.MsgSubmitEvidence{}})
	nestedAuthzMsgExecSlashing := authz.NewMsgExec(
		sdk.AccAddress{},
		[]sdk.Msg{&authzMsgExecEvidence},
	)
```

File: app/consumer/ante/disabled_modules_ante.go [hunk 3]
```diff
@@ -20,12 +21,52 @@ func (dmd DisabledModulesDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simul
...
+	}
+
+	sdkMsgs, err := msgExec.GetMessages()
+	if err != nil {
+		return []sdk.Msg{}
+	}
+
+	wrappedMsgs = append(wrappedMsgs, sdkMsgs...)
+
+	return wrappedMsgs
```
[the remark this answers is not in the document]
Comment (sontrinh16) [reply app/consumer/ante/disabled_modules_ante.go:59]: oh yeah, missed this one, since last time this func loop through msg array

**Comment by sontrinh16** [comment]
> Thanks @sontrinh16. See my comments bellow (nits). My only concern is about nested authz exec messages. For example, the following message should be filtered out, but it is not:
> 
> ```go
> 	authzMsgExecEvidence := authz.NewMsgExec(sdk.AccAddress{}, []sdk.Msg{&evidencetypes.MsgSubmitEvidence{}})
> 	nestedAuthzMsgExecSlashing := authz.NewMsgExec(
> 		sdk.AccAddress{},
> 		[]sdk.Msg{&authzMsgExecEvidence},
> 	)
> ```

hey i have added logic to cover the nested case

**Comment by MSalopek** [comment]
> I think this change requires a coordinated upgrade, right? Otherwise, some nodes may filter out the newly disabled types, while other nodes would still process them. Does that seem right to you?

No, everything in `/app` is for our simapps/testing. We recommend that the consumer teams implement these methods following our examples. 

Only the modules in `x/*` are actually importable library code.
 
Do you feel that is being communicated clearly?

**Comment by p-offtermatt** [comment]
> > I think this change requires a coordinated upgrade, right? Otherwise, some nodes may filter out the newly disabled types, while other nodes would still process them. Does that seem right to you?
> 
> No, everything in `/app` is for our simapps/testing. We recommend that the consumer teams implement these methods following our examples.
> 
> Only the modules in `x/*` are actually importable library code.
> 
> Do you feel that is being communicated clearly?

Makes sense to me - I didn't pay enough attention to where this file was, I think, but this is subtle, so I don't think we should worry too much about it in this particular case.

**Review by MSalopek — APPROVED** [review]
Thanks for taking this on!

LGTM

**Review by mpoke — APPROVED** [review]
LGTM. Thanks @sontrinh16

**Pull request merged**